### PR TITLE
[MRG] Change listener to write incoming objects in hierarchy

### PR DESCRIPTION
**Pull request by sjswerdloff**
hierarchy of patient/study/series/instance
with PatientID for patient directory name
Study and Series Instance UID respectively for study and series directory

this matches the DICOM Q/R hierarchy used for navigational queries (simplifies adding Q/R capabilities with the same hierarchy as the backing store).
It is also useful for applications that look for data in this hierarchy (e.g. DICOMpyler, OnkoDICOM)

### CODE REVIEW

File: lib/pymedphys/_dicom/connect/listen.py
Comment (SimonBiggs) [inline lib/pymedphys/_dicom/connect/listen.py:114]: Nit.

Is this function better separated from out of the class? It seems like it might be generally useful as a part of a "DICOM storage" part of PyMedPhys. A standalone function that receives a directory and a DICOM file and it spits out the PATH to where that DICOM file should be found/placed.

File: lib/pymedphys/_dicom/connect/listen.py
Comment (SimonBiggs) [inline lib/pymedphys/_dicom/connect/listen.py:126]: Nit.

Potential to use pathlib directly for consistency?

```suggestion
        series_dir.mkdir(exist_ok=True, parents=True)
```

File: lib/pymedphys/tests/dicom/test_connect.py
Comment (SimonBiggs) [inline lib/pymedphys/tests/dicom/test_connect.py:271]: Nit.

Move this function into the library and utilise it within the class (as discussed above)? Then, the `hierarchical_dicom_storage_directory` function can be tested in a standalone fashion with something like:


```python
patient_id = '000'
uid_a = pydicom.uid.generate_uid()
uid_b = pydicom.uid.generate_uid()

dataset = dicom_dataset_from_dict({
	'PatientID': patient_id ,
	'StudyInstanceUID': uid_a,
    'SeriesInstanceUID': uid_b
})

root_dir = pathlib.Path('a/directory')
expected_directory = root_dir.joinpath(patient_id, uid_a, uid_b)

created_directory = hierarchical_dicom_storage_directory(root_dir, dataset)
assert created_directory == expected_directory
```

**Review by SimonBiggs — APPROVED** [review]
Hi @sjswerdloff,

Thanks for this :) Nice manageable bite sized PR (much more manageable than my own :/). All comments here are nitpicks :). Has my approval :). Happy for you to merge once you're happy with it. Given @pchlap is the designated code owner here it would be worth waiting for his review though.

Cheers,
Simon

**Comment by sjswerdloff** [comment]
@SimonBiggs I will update to address the nits except maybe extracting the hierarchical_dicom_storage_directory.  I'll look to see if there is a good module for it to live in, but if not, I'd rather wait until the *next* thing that needs it.
And I don't intend to merge until I hear from @pchlap

**Comment by SimonBiggs** [comment]
> I'd rather wait until the _next_ thing that needs it.

Yup good call. I certainly have a preference to have functions that could be standalone to not be bundled in with a class. Even just in the same file. But it is a preference.

File: lib/pymedphys/_dicom/connect/listen.py
Comment (pchlap) [reply lib/pymedphys/_dicom/connect/listen.py:114]: This can be addressed when this is refactored out in the future, but perhaps rename the variable from `test_dataset` to `dicom_dataset` or `dataset`.

**Review by pchlap — APPROVED** [review]
Thank you @sjswerdloff 
I don't have much to add to @SimonBiggs comments. It looks good to me and am fine with refactoring out that one function in the future when it is needed. Happy for you to merge this whenever you're ready.

**Comment by sjswerdloff** [comment]
> perhaps rename the variable from test_dataset to dicom_dataset or dataset.

Yup.  Good idea.  I'll address this as well.
I intend to get these done by tomorrow evening.

**Comment by SimonBiggs** [comment]
LGTM :)

**Comment by sjswerdloff** [comment]
> LGTM :)

Great.  I'm going to wait for two more hours or all tests to complete, whichever comes first, and then I'll squash and merge

**Comment by sjswerdloff** [comment]
> > LGTM :)
> 
> Great. I'm going to wait for two more hours or all tests to complete, whichever comes first, and then I'll squash and merge

and of course it completed the moment I posted that...

**Comment by sjswerdloff** [comment]
@pchlap and @SimonBiggs , just used this for some exploratory work for OnkoDICOM.  Worked very nicely!

**Pull request merged**
